narwhal/consensus/src/consensus.rs

```diff
@@ -131,16 +131,22 @@ impl ConsensusState {
                 "Ignoring certificate {:?} as it is at or before last committed round {} for this origin",
                 certificate, origin_last_committed_round
             );
-            return false;
+            return Ok(false);
         }
 
-        dag.entry(certificate.round())
-            .or_default()
-            .insert(
-                certificate.origin(),
-                (certificate.digest(), certificate.clone()),
-            )
-            .is_none()
+        if let Some((_, existing_certificate)) = dag.entry(certificate.round()).or_default().insert(
+            certificate.origin(),
+            (certificate.digest(), certificate.clone()),
+        ) {
+            // we want to error only if we try to insert a different certificate in the dag
+            if existing_certificate != certificate.clone() {
```

Comment:
Can we just compare digest?

Reply:
We could , I just used certificate comparison as it's implementing the `Eq` trait . But sure we can compare digests